File: pkg/auth/rbac.go
package auth

import (
	"fmt"
	"sync"
)

// Role represents a user role
type Role string

const (
	RoleAdmin     Role = "admin"
	RoleUser      Role = "user"
	RoleDeveloper Role = "developer"
	RoleGuest     Role = "guest"
)

// Permission represents a permission
type Permission string

const (
	// User permissions
	PermissionReadUser   Permission = "user:read"
	PermissionWriteUser  Permission = "user:write"
	PermissionDeleteUser Permission = "user:delete"

	// Chat permissions
	PermissionChat        Permission = "chat:use"
	PermissionChatHistory Permission = "chat:history"

	// Knowledge Graph permissions
	PermissionReadKG   Permission = "kg:read"
	PermissionWriteKG  Permission = "kg:write"
	PermissionDeleteKG Permission = "kg:delete"

	// RAG permissions
	PermissionUseRAG    Permission = "rag:use"
	PermissionManageRAG Permission = "rag:manage"

	// Agent permissions
	PermissionUseAgent    Permission = "agent:use"
	PermissionManageAgent Permission = "agent:manage"

	// System permissions
	PermissionReadMetrics  Permission = "system:metrics"
	PermissionManageSystem Permission = "system:manage"
)

// RBACManager manages role-based access control
type RBACManager struct {
	mu              sync.RWMutex
	rolePermissions map[Role][]Permission
}

// NewRBACManager creates a new RBAC manager
func NewRBACManager() *RBACManager {
	manager := &RBACManager{
		rolePermissions: make(map[Role][]Permission),
	}
	manager.initializeDefaultRoles()
	return manager
}

// initializeDefaultRoles initializes default role permissions
func (m *RBACManager) initializeDefaultRoles() {
	// Admin - Full access
	m.rolePermissions[RoleAdmin] = []Permission{
		PermissionReadUser,
		PermissionWriteUser,
		PermissionDeleteUser,
		PermissionChat,
		PermissionChatHistory,
		PermissionReadKG,
		PermissionWriteKG,
		PermissionDeleteKG,
		PermissionUseRAG,
		PermissionManageRAG,
		PermissionUseAgent,
		PermissionManageAgent,
		PermissionReadMetrics,
		PermissionManageSystem,
	}

	// Developer - Development access
	m.rolePermissions[RoleDeveloper] = []Permission{
		PermissionReadUser,
		PermissionChat,
		PermissionChatHistory,
		PermissionReadKG,
		PermissionWriteKG,
		PermissionUseRAG,
		PermissionManageRAG,
		PermissionUseAgent,
		PermissionManageAgent,
		PermissionReadMetrics,
	}

	// User - Standard user access
	m.rolePermissions[RoleUser] = []Permission{
		PermissionReadUser,
		PermissionChat,
		PermissionChatHistory,
		PermissionReadKG,
		PermissionUseRAG,
		PermissionUseAgent,
	}

	// Guest - Limited access
	m.rolePermissions[RoleGuest] = []Permission{
		PermissionChat,
		PermissionReadKG,
	}
}

// HasPermission checks if a role has a specific permission
func (m *RBACManager) HasPermission(role Role, permission Permission) bool {
	m.mu.RLock()
	defer m.mu.RUnlock()

	permissions, exists := m.rolePermissions[role]
	if !exists {
		return false
	}

	for _, p := range permissions {
		if p == permission {
			return true
		}
	}
	return false
}

// HasAnyPermission checks if a role has any of the specified permissions
func (m *RBACManager) HasAnyPermission(role Role, permissions ...Permission) bool {
	for _, permission := range permissions {
		if m.HasPermission(role, permission) {
			return true
		}
	}
	return false
}

// HasAllPermissions checks if a role has all of the specified permissions
func (m *RBACManager) HasAllPermissions(role Role, permissions ...Permission) bool {
	for _, permission := range permissions {
		if !m.HasPermission(role, permission) {
			return false
		}
	}
	return true
}

// CheckUserPermission checks if a user (with multiple roles) has a permission
func (m *RBACManager) CheckUserPermission(roles []string, permission Permission) bool {
	for _, roleStr := range roles {
		role := Role(roleStr)
		if m.HasPermission(role, permission) {
			return true
		}
	}
	return false
}

// CheckUserPermissions checks if a user has all specified permissions
func (m *RBACManager) CheckUserPermissions(roles []string, permissions ...Permission) bool {
	for _, permission := range permissions {
		hasPermission := false
		for _, roleStr := range roles {
			role := Role(roleStr)
			if m.HasPermission(role, permission) {
				hasPermission = true
				break
			}
		}
		if !hasPermission {
			return false
		}
	}
	return true
}

// AddPermissionToRole adds a permission to a role
func (m *RBACManager) AddPermissionToRole(role Role, permission Permission) {
	m.mu.Lock()
	defer m.mu.Unlock()

	if permissions, exists := m.rolePermissions[role]; exists {
		// Check if permission already exists
		for _, p := range permissions {
			if p == permission {
				return
			}
		}
		m.rolePermissions[role] = append(permissions, permission)
	} else {
		m.rolePermissions[role] = []Permission{permission}
	}
}

// RemovePermissionFromRole removes a permission from a role
func (m *RBACManager) RemovePermissionFromRole(role Role, permission Permission) {
	m.mu.Lock()
	defer m.mu.Unlock()

	if permissions, exists := m.rolePermissions[role]; exists {
		newPermissions := make([]Permission, 0, len(permissions))
		for _, p := range permissions {
			if p != permission {
				newPermissions = append(newPermissions, p)
			}
		}
		m.rolePermissions[role] = newPermissions
	}
}

// GetRolePermissions returns all permissions for a role
func (m *RBACManager) GetRolePermissions(role Role) []Permission {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if permissions, exists := m.rolePermissions[role]; exists {
		// Return a copy to prevent external modification
		result := make([]Permission, len(permissions))
		copy(result, permissions)
		return result
	}
	return []Permission{}
}

// ValidateRole checks if a role is valid
func (m *RBACManager) ValidateRole(role string) error {
	m.mu.RLock()
	defer m.mu.RUnlock()

	r := Role(role)
	if _, exists := m.rolePermissions[r]; !exists {
		return fmt.Errorf("invalid role: %s", role)
	}
	return nil
}

File: algo/agent-engine/app/auth/rbac.py
"""
RBAC (Role-Based Access Control) for Python services
"""

from enum import Enum
from typing import Dict, List, Set


class Role(str, Enum):
    """User roles"""

    ADMIN = "admin"
    USER = "user"
    DEVELOPER = "developer"
    GUEST = "guest"


class Permission(str, Enum):
    """Permissions"""

    # User permissions
    READ_USER = "user:read"
    WRITE_USER = "user:write"
    DELETE_USER = "user:delete"

    # Chat permissions
    CHAT = "chat:use"
    CHAT_HISTORY = "chat:history"

    # Knowledge Graph permissions
    READ_KG = "kg:read"
    WRITE_KG = "kg:write"
    DELETE_KG = "kg:delete"

    # RAG permissions
    USE_RAG = "rag:use"
    MANAGE_RAG = "rag:manage"

    # Agent permissions
    USE_AGENT = "agent:use"
    MANAGE_AGENT = "agent:manage"

    # System permissions
    READ_METRICS = "system:metrics"
    MANAGE_SYSTEM = "system:manage"


class RBACManager:
    """RBAC manager"""

    def __init__(self):
        self.role_permissions: Dict[Role, Set[Permission]] = {}
        self._initialize_default_roles()

    def _initialize_default_roles(self):
        """Initialize default role permissions"""
        # Admin - Full access
        self.role_permissions[Role.ADMIN] = {
            Permission.READ_USER,
            Permission.WRITE_USER,
            Permission.DELETE_USER,
            Permission.CHAT,
            Permission.CHAT_HISTORY,
            Permission.READ_KG,
            Permission.WRITE_KG,
            Permission.DELETE_KG,
            Permission.USE_RAG,
            Permission.MANAGE_RAG,
            Permission.USE_AGENT,
            Permission.MANAGE_AGENT,
            Permission.READ_METRICS,
            Permission.MANAGE_SYSTEM,
        }

        # Developer - Development access
        self.role_permissions[Role.DEVELOPER] = {
            Permission.READ_USER,
            Permission.CHAT,
            Permission.CHAT_HISTORY,
            Permission.READ_KG,
            Permission.WRITE_KG,
            Permission.USE_RAG,
            Permission.MANAGE_RAG,
            Permission.USE_AGENT,
            Permission.MANAGE_AGENT,
            Permission.READ_METRICS,
        }

        # User - Standard user access
        self.role_permissions[Role.USER] = {
            Permission.READ_USER,
            Permission.CHAT,
            Permission.CHAT_HISTORY,
            Permission.READ_KG,
            Permission.USE_RAG,
            Permission.USE_AGENT,
        }

        # Guest - Limited access
        self.role_permissions[Role.GUEST] = {
            Permission.CHAT,
            Permission.READ_KG,
        }

    def has_permission(self, role: Role, permission: Permission) -> bool:
        """Check if a role has a specific permission"""
        return permission in self.role_permissions.get(role, set())

    def has_any_permission(
        self, role: Role, permissions: List[Permission]
    ) -> bool:
        """Check if a role has any of the specified permissions"""
        role_perms = self.role_permissions.get(role, set())
        return any(perm in role_perms for perm in permissions)

    def has_all_permissions(
        self, role: Role, permissions: List[Permission]
    ) -> bool:
        """Check if a role has all of the specified permissions"""
        role_perms = self.role_permissions.get(role, set())
        return all(perm in role_perms for perm in permissions)

    def check_user_permission(
        self, roles: List[str], permission: Permission
    ) -> bool:
        """Check if a user (with multiple roles) has a permission"""
        for role_str in roles:
            try:
                role = Role(role_str)
                if self.has_permission(role, permission):
                    return True
            except ValueError:
                continue
        return False

    def check_user_permissions(
        self, roles: List[str], permissions: List[Permission]
    ) -> bool:
        """Check if a user has all specified permissions"""
        for permission in permissions:
            if not self.check_user_permission(roles, permission):
                return False
        return True

    def add_permission_to_role(self, role: Role, permission: Permission):
        """Add a permission to a role"""
        if role not in self.role_permissions:
            self.role_permissions[role] = set()
        self.role_permissions[role].add(permission)

    def remove_permission_from_role(self, role: Role, permission: Permission):
        """Remove a permission from a role"""
        if role in self.role_permissions:
            self.role_permissions[role].discard(permission)

    def get_role_permissions(self, role: Role) -> List[Permission]:
        """Get all permissions for a role"""
        return list(self.role_permissions.get(role, set()))

    def validate_role(self, role: str) -> bool:
        """Validate if a role is valid"""
        try:
            Role(role)
            return True
        except ValueError:
            return False


# Global RBAC manager instance
_rbac_manager: RBACManager = None


def get_rbac_manager() -> RBACManager:
    """Get the global RBAC manager instance"""
    global _rbac_manager
    if _rbac_manager is None:
        _rbac_manager = RBACManager()
    return _rbac_manager
